examples/silentpayments.c

```diff
@@ -0,0 +1,417 @@
+            assert(ret);
+            recipients[i].index = i;
+            recipient_ptrs[i] = &recipients[i];
+        }
+        for (i = 0; i < N_TX_OUTPUTS; i++) {
+            generated_output_ptrs[i] = &tx_outputs[i];
+        }
+        ret = secp256k1_silentpayments_sender_create_outputs(ctx,
+            generated_output_ptrs,
+            recipient_ptrs, N_TX_OUTPUTS,
+            smallest_outpoint,
+            sender_seckey_ptrs, N_TX_INPUTS,
+            NULL, 0
+        );
+        assert(ret);
+        printf("Alice created the following outputs for Bob and Carol: \n");
+        for (i = 0; i < N_TX_OUTPUTS; i++) {
+            printf("    ");
+            printf("%s : ", address_amounts[i]);
+            secp256k1_xonly_pubkey_serialize(ctx, xonly_print, &tx_outputs[i]);
+            print_hex(xonly_print, sizeof(xonly_print));
+        }
+        /* It's best practice to try to clear secrets from memory after using them.
+         * This is done because some bugs can allow an attacker to leak memory, for
+         * example through "out of bounds" array access (see Heartbleed), Or the OS
+         * swapping them to disk. Hence, we overwrite the secret key buffer with zeros.
+         *
+         * Here we are preventing these writes from being optimized out, as any good compiler
+         * will remove any writes that aren't used. */
+        for (i = 0; i < N_TX_INPUTS; i++) {
+            secure_erase(&sender_seckeys[i], sizeof(sender_seckeys[i]));
+        }
+    }
+
+    /*** Receiving ***/
+    {
+        /*** Transaction data ***
+         *
+         * Here we create a few global variables to represent the transaction data:
+         *
+         *     1. The transaction inputs, `tx_input_ptrs`
+         *     2. The transaction outputs, `tx_output_ptrs`
+         *
+         * These will be used to demonstrate scanning as a full node and scanning as a light client.
+         */
+        const secp256k1_xonly_pubkey *tx_input_ptrs[N_TX_INPUTS];
+        const secp256k1_xonly_pubkey *tx_output_ptrs[N_TX_OUTPUTS];
+        unsigned char light_client_data33[33];
+
+        for (i = 0; i < N_TX_INPUTS; i++) {
+            tx_input_ptrs[i] = &tx_inputs[i];
+        }
+        for (i = 0; i < N_TX_OUTPUTS; i++) {
+            tx_output_ptrs[i] = &tx_outputs[i];
+        }
+
+        /*** Scanning with labels as a full node (Bob) ***
+         *
+         * Since Bob has access to the full transaction, scanning is simple:
+         *
+         *     1. Collect the relevant data from the transaction inputs and call
+         *        `secp256k1_silentpayments_recipient_public_data_create`
+         *     2. Call `secp256k1_silentpayments_recipient_scan_outputs`
+         *
+         * Additionally, since Bob has access to the full transaction outputs when scanning its easy for him
+         * to scan with labels, as demonstrated below. For efficient scanning, Bob keeps a cache of
```

Comment:
c71791c629ec88e2e253aae4157d7113b60bee60: the example would be easier to follow if you first demonstrate usage without labels, since they're optional.

Reply:
I considered this, but as mentioned in https://github.com/bitcoin-core/secp256k1/pull/1519#discussion_r1673844535, the goal of the example is to provide guidance for the more complicated usage of the API. In this case, the goal is to give a concrete example of what the `label_context` is.

Perhaps its worth adding a comment mentioning that labels are optional and using without labels is as simple as calling the function with `NULL, NULL` for label_lookup and label_context.

Reply:
Same reaction. I think you need to guide people through a simple usage before covering a more advanced one. Otherwise anyone just looking for the simple use case will give up confused. And most people looking for an advanced use case probably want to start with the simple one and then expand it.

It's already clear from the comment that labels are optional, but it's not clear how to avoid using them (without staring for a long time). You could perhaps add comments along the lines of "// Skip this if you're not using labels".

Reply:
I'd prefer to keep labels in the example since it shows usage of the `secp256k1_silentpayments_recipient_create_label_tweak` function. I added a comment near the top that makes it more clear that labels are optional and that the function can be called with `NULL, NULL` if not using labels. This comment is then reiterated in the scanning section with a comment at the call site of the scanning function.